File: exam2/pasta/MinHeap.java
public class MinHeap {
    private int[] A;
    private int size;

    public void buildMinHeap(int[] arr) {
        size = arr.length;
        this.A = new int[size + 1];

        for (int i = 1; i <= size; i++) {
            this.A[i] = arr[i - 1];
        }
        for (int i = (int) Math.floor(size / 2); i >= 1; i--) {
            minHeapify(i);
        }
    }

    public void minHeapify(int i) {
        int l = getLeft(i);
        int r = getRight(i);
        int smallest;

        if (l <= size && A[l] < A[i]) {
            smallest = l;
        } else {
            smallest = i;
        }

        if (r <= size && A[r] < A[smallest]) {
            smallest = r;
        }

        if (smallest != i) {
            swap(i, smallest);
            minHeapify(smallest);
        }
    }

    public void extractMin() {
        if (size < 1) {
            System.out.print("Error: heap empty\n");
        } else {
            int min = A[1];
            A[1] = A[size];
            size--;
            minHeapify(1);
        }
    }

    public void insert(int key) {
        size++;
        int[] temp = new int[size + 1];
        for (int i = 1; i <= size - 1; i++) {
            temp[i] = A[i];
        }
        A = temp;
        decreaseKey(size, key);
    }

    public void decreaseKey(int i, int key) {
        if (key > A[i] && A[i] != 0) {
            System.out.println("Error: new key is greater than current key");
        }
        A[i] = key;
        while (i > 1 & A[getParent(i)] > A[i]) {
            swap(i, getParent(i));
            i = getParent(i);
        }
    }

    public void heapSort() {
        buildMinHeap(A);
        for (int i = A.length; i >= 2; i--) {
            swap(1, i);
            size--;
            minHeapify(1);
        }
    }

    public void swap(int a, int b) {
        int temp = A[a];
        A[a] = A[b];
        A[b] = temp;
    }

    private int getParent(int index) {
        return (int) Math.floor(index / 2);
    }

    private int getLeft(int index) {
        return (2 * index);
    }

    private int getRight(int index) {
        return (2 * index + 1);
    }

    public void printArr() {
        System.out.print("i:\t");
        for (int x = 1; x <= size; x++) {
            System.out.print("\t" + x);
        }
        System.out.print("\nA[i]:\t");
        for (int x = 1; x <= size; x++) {
            System.out.print("\t" + A[x]);
        }
        System.out.print("\n");
    }
}

File: exam2/tchow/MinHeap.cpp
/* This is a MIN heap program. All you need to do is enter the array and the heap commands in main.
 * This program provides outputs similar to project 2. This program will automatically build the heap and
 * heapify the array you entered similar to project 2 aswell as provide the number of heapify calls when
 * build heap and extract are run. This heap uses indexing at 1 as you can see when the heap is outputted.
*/

// A C++ program to demonstrate common Binary Heap Operations
#include<iostream>
#include<climits>
using namespace std;

// Prototype of a utility function to swap two integers
void swap(int *x, int *y);

// A class for Min Heap
class MinHeap
{
    int *harr; // pointer to array of elements in heap
    int capacity; // maximum possible size of min heap
    int heap_size; // Current number of elements in min heap
public:
    // Constructor
    MinHeap(int capacity, int arr[],int size);

    // to heapify a subtree with the root at given index
    void MinHeapify(int ,int &counter);

    void buildHeap(int arr[],int size);

    int parent(int i) { return (i)/2; }

    // to get index of left child of node at index i
    int left(int i) { return (2*i); }

    // to get index of right child of node at index i
    int right(int i) { return (2*i+1); }

    // to extract the root which is the minimum element
    void extractMin();

    // Decreases key value of key at index i to new_val
    void decreaseKey(int i, int new_val);

    // Inserts a new key 'k'
    void insertKey(int k);

    void printHeap();

};

void MinHeap::buildHeap(int arr[],int size) {

    int counter = 0;
    for(int i=1;i<=size;i++){
        harr[i]=arr[i-1];
        heap_size++;
    }
    for (int iterate = heap_size/2;iterate >= 1;iterate--){
        MinHeapify(iterate,counter);
    }
    cout<<"The number of heapify calls from build heap: "<<counter<<"\n";
}

// Constructor: Builds a heap from a given array a[] of given size
MinHeap::MinHeap(int cap,int arr[],int size)
{
    heap_size = 0;
    capacity = cap;
    harr = new int[cap+1];
    buildHeap(arr,size);
    cout<<"This is the original min heap:\n";
    printHeap();
    cout<<"\nHeap commands:\n";
}

// Inserts a new key 'k'
void MinHeap::insertKey(int k)
{
    cout<<"InsertKey("<<k<<")\n";

    if (heap_size == capacity)
    {
        cout << "\nOverflow: Could not insertKey\n";
        return;
    }

    // First insert the new key at the end
    heap_size++;
    int i = heap_size;

    // Fix the min heap property if it is violated
    while (i > 1 && harr[parent(i)] > k)
    {
        harr[i] = harr[parent(i)];
        i = parent(i);
    }
    harr[i] = k;
    cout<<"After insert:\n";
    printHeap();
}

// Decreases value of key at index 'i' to new_val.  It is assumed that
// new_val is smaller than harr[i].
void MinHeap::decreaseKey(int i, int new_val)
{
    
     if(new_val>harr[i]){
        cout<<"Error: new key is larger than current\n";
        printHeap();
        return;
    }
    cout<<"decreaseKey("<<i<<","<<new_val<<")\n";
    harr[i] = new_val;
    while (i > 1 && harr[parent(i)] > harr[i])
    {
        swap(&harr[i], &harr[parent(i)]);
        i = parent(i);
    }
    cout<<"After decreaseKey:\n";
    printHeap();
}

// Method to remove minimum element (or root) from min heap
void MinHeap::extractMin()
{
    cout<<"extractMin()\n";
    int counter = 0;
    if (heap_size < 1){
        cout<<"Heap is empty \n";
        return;
    }
    // Store the minimum value, and remove it from heap
    int root = harr[1];
    harr[1] = harr[heap_size];
    heap_size--;
    MinHeapify(1,counter);
    cout<<"The number of heapify calls from extractMin: "<<counter<<"\n";
    cout<<"Deleted key: "<<root<<"\n";
    printHeap();
}

void MinHeap::printHeap() {
    std::cout<<"capacity="<<capacity;
    std::cout<<", size="<<heap_size<<"\nIndex:\t\t";
    for(int increment = 1; increment<=heap_size;increment++){
        cout<<increment<<"\t";
    }
    std::cout<<"\nElement:\t";
    for(int increment = 1; increment<=heap_size;increment++){
        cout<<harr[increment]<<"\t";
    }
    std::cout<<"\n\n";
}

// A recursive method to heapify a subtree with the root at given index
// This method assumes that the subtrees are already heapified
void MinHeap::MinHeapify(int i, int &counter)
{
    counter++;
    int l = left(i);
    int r = right(i);
    int smallest = i;
    if (l <= heap_size && harr[l] < harr[i])
        smallest = l;
    if (r <= heap_size && harr[r] < harr[smallest])
        smallest = r;
    if (smallest != i)
    {
        swap(&harr[i], &harr[smallest]);
        MinHeapify(smallest,counter);
    }
}

// A utility function to swap two elements
void swap(int *x, int *y)
{
    int temp = *x;
    *x = *y;
    *y = temp;
}

// Driver program to test above functions
int main()
{
    //************************************************************************
    //Enter array here

    //Examples
    //int arr[] = {70,60,50,40,30,20,10};
    int arr[] = {11,13,15,17,19,21,23,25,27,29};
    
    //************************************************************************
    int size=sizeof(arr)/sizeof(arr[0]);
    MinHeap h(20,arr,size);

    //************************************************************************
    //Enter commands here
    //Format should be: h.[enter function];

    /*Examples
     * h.insertKey(4);
     * h.extractMin();
     * h.decreaseKey(3, 28);
    */
    //************************************************************************

    return 0;
}
